File: leetcode/7 - Trees/98.ValidateBinarySearchTree.ts
/* 
Given the root of a binary tree, determine if it is a valid binary search tree (BST).

A valid BST is defined as follows:

The left subtree of a node contains only nodes with keys less than the node's key.
The right subtree of a node contains only nodes with keys greater than the node's key.
Both the left and right subtrees must also be binary search trees.
*/

/* ------------------------------------------------------------------------------- */

class TreeNode {
    val: number;
    left: TreeNode | null;
    right: TreeNode | null;
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = val === undefined ? 0 : val;
        this.left = left === undefined ? null : left;
        this.right = right === undefined ? null : right;
    }
}

//DFS
// function isValidBST(root: TreeNode | null): boolean {
//     function dfs(root: TreeNode | null, max: number, min: number) {
//         if (!root) return true;
//         if (root.val <= max || root.val >= min) return false;

//         return dfs(root.left, max, root.val) && dfs(root.right, root.val, min);
//     }

//     return dfs(root, -Infinity, Infinity);
// }

/*
T.C.: O(N)
S.C.: O(H)
*/

/* ------------------------------------------------------------------------------- */

//DFS 2
function isValidBST(root: TreeNode | null): boolean {
    if (!root) return false;

    const values: number[] = [];
    function dfs(root: TreeNode) {
        root?.left && dfs(root.left);
        values.push(root.val);
        root?.right && dfs(root.right);
    }

    dfs(root);

    for (let i = 0; i < values.length; i++) {
        if (values[i] >= values[i + 1]) return false;
    }

    return true;
}

/*
T.C.: O(N)
S.C.: O(H)
*/

/* ------------------------------------------------------------------------------- */

File: python/7 - Trees/98.ValidateBinarySearchTree.py
# Definition for a binary tree node.
from collections import deque
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# class Solution:
#     def isValidBST(self, root: Optional[TreeNode]) -> bool:
#         if not root:
#             return True

#         def dfs(node: TreeNode | None, max: int, min: int):
#             if not node:
#                 return True
#             if node.val <= min or node.val >= max:
#                 return False
#             return dfs(node.left, node.val, min) and dfs(node.right, max, node.val)

#         return dfs(root, float("inf"), float("-inf"))


"""
T.C.: O(N)
 • 每個節點只會被拜訪一次，N 為樹的節點數。
 • 每次拜訪只做常數次比較和遞迴呼叫。

S.C.: O(N)
 • 主要來自遞迴呼叫堆疊。
 • 最壞情況（樹退化成一條鏈）：O(N)
 • 平衡樹時，遞迴深度為 O(logN)
"""


class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True

        q = deque([(root, float("-inf"), float("inf"))])

        while q:
            node, min, max = q.popleft()
            if node.val <= min or node.val >= max:
                return False
            if node.left:
                q.append((node.left, min, node.val))
            if node.right:
                q.append((node.right, node.val, max))
        return True


"""
T.C.: O(N)
 • 每個節點只會被放進 queue 一次、處理一次，N 為樹的節點數。
 • 每次操作（比較、入隊、出隊）都是 O(1)。

S.C.: O(N)
 • 主要來自 queue（佇列）的儲存。
 • 最壞情況（完全不平衡的樹或完全二元樹的最後一層），queue 可能同時存放 O(N) 個節點。
"""
